cardano-node/src/Cardano/Node/Tracing/Tracers/BlockReplayProgress.hs

```diff
@@ -71,28 +88,34 @@ instance MetaTrace ReplayBlockStats where
 
 withReplayedBlock :: Trace IO ReplayBlockStats
     -> IO (Trace IO (ChainDB.TraceEvent blk))
-withReplayedBlock tr =
-    let tr' = filterTrace filterFunction tr
-        tr'' = contramap unfold tr'
-    in foldTraceM replayBlockStats emptyReplayBlockStats tr''
+withReplayedBlock tr = do
+  var <- newMVar initialReplayBlockState
+  contramapMCond tr (process var)
   where
-    filterFunction(_, ReplayBlockStats {..}) = rpsDisplay
-
-replayBlockStats :: MonadIO m
-  => ReplayBlockStats
-  -> LoggingContext
-  -> ChainDB.TraceEvent blk
-  -> m ReplayBlockStats
-replayBlockStats ReplayBlockStats {..} _context
-    (ChainDB.TraceLedgerDBEvent
-     (LedgerDB.LedgerReplayEvent
-      (LedgerDB.TraceReplayProgressEvent
-       (LedgerDB.ReplayedBlock pt [] _ (LedgerDB.ReplayGoal replayTo))))) = do
-      let slotno = realPointSlot pt
-          endslot = withOrigin 0 id $ pointSlot replayTo
-          progress' = (fromIntegral (unSlotNo slotno) * 100.0) / fromIntegral (unSlotNo $ max slotno endslot)
-      pure $ if (progress' == 0.0 && not rpsDisplay)
-                || ((progress' - rpsLastProgress) > 0.1)
-                then ReplayBlockStats True slotno endslot progress' progress'
-                else ReplayBlockStats False slotno endslot progress' rpsLastProgress
-replayBlockStats st@ReplayBlockStats {} _context _ = pure st
+    process :: MVar ReplayBlockState
+            -> (LoggingContext, Either TraceControl (ChainDB.TraceEvent blk))
+            -> IO (Maybe (LoggingContext, Either TraceControl ReplayBlockStats))
+    process _ (ctx, Left control) = pure (Just (ctx, Left control))
+    process var (ctx, Right msg) = modifyMVar var $ \st -> do
+      let (st', mbStats) = mbProduceBlockStats st msg
+      pure (st', fmap ((ctx,) . Right) mbStats)
+
+    mbProduceBlockStats :: ReplayBlockState -> ChainDB.TraceEvent blk -> (ReplayBlockState, Maybe ReplayBlockStats)
```

Comment:
You could reduce code duplication and improve readability by changing the type to
```
mbProduceBlockStats :: ChainDB.TraceEvent blk -> ReplayBlockState -> (ReplayBlockState, Maybe ReplayBlockStats)
```

Like this, you need to match on the event only once, and then add a conditional guard like (pseudo-code):
```
arg: st@(ReplayBlockState rbs)
   ...
    < match >
       |  maybe True (\lastSlot -> progressFor etc.) rbs -> emit and update state
    _ -> (st, Nothing)
```

Like this, it becomes much better readable under which condition the trace is emitted, and also it's clear that there's only 1 code path so the trace emitted will be identical.